### ISSUE TITLE
Object pose estimation with BPnP

### ISSUE THREAD

**Issue by tzole1155** [BoChenYS/BPnP#10]
Hi Bo,

We are trying to reproduce your results presented in **Section 5** of the paper so we would like to ask for some clarifications.

- Could you share some details about the configuration of _HRNet_ (e.g. nr of stages, heatmaps resolution, etc.) and the training parameters (e.g. learning rate, optimizer, batch size, etc.) 

- Regarding the heatmap loss have you tried others expect _MSE_ (e.g. a _KL_ loss instead)?. We've noticed better convergence with an _L1_ loss so far.

- Have you added any priors to the predicted heatmaps (e.g. softmax). Additionally, the gt heatmaps are in [0,W] x [0, H] format or normalised?

Thanks for your time.

Best,

Giorgos

**Comment by BoChenYS** [comment]
Hi Giorgos,

The HRNet I used had this architecture:

MODEL:
  INIT_WEIGHTS: true
  PRETRAINED: 'pretrained_hrnet/hrnet_w32-36af842e.pth'
  TARGET_TYPE: gaussian
  IMAGE_SIZE:
  - 640
  - 480
  EXTRA:
    PRETRAINED_LAYERS:
    - 'conv1'
    - 'bn1'
    - 'conv2'
    - 'bn2'
    - 'layer1'
    - 'transition1'
    - 'stage2'
    - 'transition2'
    - 'stage3'
    - 'transition3'
    - 'stage4'
    FINAL_CONV_KERNEL: 1
    STAGE2:
      NUM_MODULES: 1
      NUM_BRANCHES: 2
      BLOCK: BASIC
      NUM_BLOCKS:
      - 4
      - 4
      NUM_CHANNELS:
      - 32
      - 64
      FUSE_METHOD: SUM
    STAGE3:
      NUM_MODULES: 4
      NUM_BRANCHES: 3
      BLOCK: BASIC
      NUM_BLOCKS:
      - 4
      - 4
      - 4
      NUM_CHANNELS:
      - 32
      - 64
      - 128
      FUSE_METHOD: SUM
    STAGE4:
      NUM_MODULES: 3
      NUM_BRANCHES: 4
      BLOCK: BASIC
      NUM_BLOCKS:
      - 4
      - 4
      - 4
      - 4
      NUM_CHANNELS:
      - 32
      - 64
      - 128
      - 256
      FUSE_METHOD: SUM


The heatmap size is one fourth of the the input image size, i.e., in the LINEMOD case, 120*160 resolution. Initial learning rate is 0.001 with Adam optimizer and 0.1 learning rate decay at milestones=[50, 90, 110] epochs. Batch size is 16 (but it should depend on your GPU memory). 

I haven't tried KL loss for the heatmap but I think it should work too. Haven't tried L1 either. Interesting to hear that L1 converges better...

Yes heatmaps are flat_softmax normalised, that is, solftmax over all pixels so that each channel sums to 1.  The gt heatmaps are not coordinates, gt heatmaps are heatmaps, which are 2D Gaussians maps (120*160 resolution). (so it doesn't matter whether the gt coordinates are in [0,W]*[0, H] or normalised in [0,1]*[0, 1], right?) 

Hope that helps. 

Cheers
Bo

**Comment by tzole1155** [comment]
Hi Bo,

Thanks for the detailed answer. 

Sorry for the misunderstanding I haven't formulated this correctly. I was reffering to the predicted coordinateds.
 
Thanks,

Giorgos
